8308184: Launching java with large number of jars in classpath with java.protocol.handler.pkgs system property set can lead to StackOverflowError

srutjay opened the pull request:
This is the backport of the issue : [JDK-8308184](https://bugs.openjdk.org/browse/JDK-8308184)

Launching java with large number of jars in classpath with java.protocol.handler.pkgs system property set can lead to StackOverflowError

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8308184](https://bugs.openjdk.org/browse/JDK-8308184) needs maintainer approval

### Issue
 * [JDK-8308184](https://bugs.openjdk.org/browse/JDK-8308184): Launching java with large number of jars in classpath with java.protocol.handler.pkgs system property set can lead to StackOverflowError (**Bug** - P4 - Approved)


### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk11u-dev.git pull/2749/head:pull/2749` \
`$ git checkout pull/2749`

Update a local copy of the PR: \
`$ git checkout pull/2749` \
`$ git pull https://git.openjdk.org/jdk11u-dev.git pull/2749/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2749`

View PR using the GUI difftool: \
`$ git pr show -t 2749`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk11u-dev/pull/2749.diff">https://git.openjdk.org/jdk11u-dev/pull/2749.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk11u-dev/pull/2749#issuecomment-2155250954)

srutjay commented:
/signed

offamitkumar commented:
Please enable GHA test on your fork;

offamitkumar commented on src/java.base/share/classes/jdk/internal/loader/URLClassPath.java:
```diff
@@ -208,7 +208,9 @@ public URLClassPath(URL[] urls, AccessControlContext acc) {
         this.unopenedUrls = unopenedUrls;
         this.path = path;
-        this.jarHandler = null;
+        // the application class loader uses the built-in protocol handler to avoid protocol
```
It should have been a clean backport. Please remove these two imports from this file: 

```java
import java.util.HashSet;
import java.util.Set;
```

Also please mention about testing you have done for this patch !!

srutjay replied:
@offamitkumar 

The imports are necessary and cannot be removed because they are utilized in the code at lines 952 and 964.

Regarding the testing :
I have a sample Java program that prints "Hello, world." After compiling the code, I created two JAR files from the .class file. When I run the Java command with the created JARs in the classpath and the -Djava.protocol.handler.pkgs system property set to a package name,  when I use the -verbose option along , the second JAR in the classpath gets loaded first.

TheRealMDoerr commented:
There may be another way to make it a clean backport: Rename the title to "Backport 0a4f92603b76181358326908497220c7d14211b1" which will make it a backport of the JDK17 version.

srutjay commented:
@TheRealMDoerr I have updated the Title as suggested . Thank you for your guidance !

TheRealMDoerr commented:
Ok, the actual fix is clean, now. Please describe your test changes.

srutjay commented:
> Ok, the actual fix is clean, now. Please describe your test changes.

I added _JarBuilder.java_ because the _LargeClasspathWithPkgPrefix.java_ test case depended on _test/lib/jdk/test/lib/util/JarBuilder.java_. 

Additionally, I updated the String literal handling; since Java multi-line string literals are supported only from JDK 13, I modified the code to use a StringBuilder instead.

TheRealMDoerr commented:
Can't you use test/jdk/lib/testlibrary/java/util/jar/JarBuilder.java? It has been moved by https://github.com/openjdk/jdk/commit/dda0c6cecd38b4897c377e10ffb0ebe75360ac43.

srutjay commented:
> Can't you use test/jdk/lib/testlibrary/java/util/jar/JarBuilder.java? It has been moved by [openjdk/jdk@dda0c6c](https://github.com/openjdk/jdk/commit/dda0c6cecd38b4897c377e10ffb0ebe75360ac43).

@TheRealMDoerr I have done the suggested changes.

TheRealMDoerr commented on test/jdk/sun/misc/URLClassPath/LargeClasspathWithPkgPrefix.java:
```diff
@@ -0,0 +1,144 @@
+          .append("}\n");
+
+        JAVA_MAIN_CONTENT = sb.toString();
+}
```
Multi-line Strings are typically written like https://github.com/openjdk/jdk11u-dev/blob/f08e5e87938d3f20575ef186f4eefeb9faeb6e80/test/hotspot/jtreg/containers/cgroup/CgroupSubsystemFactory.java#L88. Please also make sure the `}` is aligned properly.

srutjay replied:
Changes done

TheRealMDoerr replied:
I suggest to add spaces within the String such that the program is better readable and looks like in the original change.

srutjay replied:
addressed

TheRealMDoerr reviewed (approved):
Thanks for the updates. Looks good, now.

srutjay commented:
/approval

srutjay commented:
/approval request This backport is necessary to resolve an issue where the second JAR file in the class path is loaded first when two JAR files are specified.

srutjay commented:
/integrate

srutjay commented:
> @srutjay Your change (at version [9fb3c54](https://github.com/openjdk/jdk11u-dev/commit/9fb3c54441be153069208f0b25aaf22e81618e93)) is now ready to be sponsored by a Committer.

@TheRealMDoerr 
Could you assist me with merging this commit?
Thank you in advance.

TheRealMDoerr commented:
/sponsor
